### PR TITLE
internal/civisibility: intelligent test runner support

### PR DESCRIPTION
**Pull request by tonyredondo**
### What does this PR do?

This PR adds the Test Visibility Intelligent Test Runner feature: https://datadoghq.atlassian.net/wiki/spaces/SDTEST/pages/3123348305/Intelligent+Test+Runner+in+CI+Visibility+Libraries

### Motivation

This is one of the most important feature for Test Visibility.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: internal/civisibility/utils/net/coverage.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// SendCoveragePayload sends a code coverage payload to the backend.
+func (c *client) SendCoveragePayload(ciTestCovPayload io.Reader) error {
+	if ciTestCovPayload == nil {
+		return errors.New("coverage payload is nil")
+	}
+
+	// Create a dummy event to send with the coverage payload.
+	dummyEvent := FormFile{
```
Comment (ManuelPalenzuelaDD) [inline internal/civisibility/utils/net/coverage.go:33]: Q: Why do we need to upload a dummy event?

File: internal/civisibility/integrations/gotesting/coverage/test_coverage.go [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+}
+
+// CollectCoverageAfterTestExecution collects coverage after test execution.
+func (t *testCoverage) CollectCoverageAfterTestExecution() {
+	if !CanCollect() {
+		return
+	}
+
+	restore := setStdOutToTemp()
+	t.postCoverageFilename = filepath.Join(temporaryDir, fmt.Sprintf("%d-%d-%d-post.out", t.moduleID, t.suiteID, t.testID))
```
Comment (ManuelPalenzuelaDD) [inline internal/civisibility/integrations/gotesting/coverage/test_coverage.go:205]: why is this line inside the `restore := setStdOutToTemp()` block? isn't it only necessary for the `tearDown` fn to be in that block?

**Review by ManuelPalenzuelaDD — APPROVED** [review]
LGTM. I lack context in a lot of the implementation details of this, but the go impl looks ok. I've left some minor Qs/nits

File: internal/civisibility/integrations/civisibility_features.go [hunk 3]
```diff
@@ -133,13 +149,26 @@ func ensureAdditionalFeaturesInitialization(serviceName string) {
...
+				utils.GetCITags()["itr_correlation_id"] = correlationId
+				ciVisibilitySkippables = skippableTests
+			}
+		}
 	})
 }
 
 // GetSettings gets the settings from the backend settings endpoint
 func GetSettings() *net.SettingsResponseData {
 	// call to ensure the additional features initialization is completed (service name can be null here)
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/civisibility_features.go]: changed in [1ba2dda](https://github.com/DataDog/dd-trace-go/pull/2943/commits/1ba2ddaeb7768e4f9cf9208d81a6db7890cdaed0)

File: internal/civisibility/integrations/gotesting/coverage/test_coverage.go (hunk 2 above)
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/coverage/test_coverage.go:209]: True, I did a refactor to remove that method and just create a tearDown wrapper in [1ba2dda](https://github.com/DataDog/dd-trace-go/pull/2943/commits/1ba2ddaeb7768e4f9cf9208d81a6db7890cdaed0)

**Pull request merged**